File: qizhi/mobile/MentorAI-iOS/MentorAI/Feedback/FeedbackModels.swift
import Foundation

/// Mirrors the server `SubmitFeedbackParams`. Star is 1–5; content is required non-empty.
/// `imagePaths` are server-side storage paths returned by `/attachment/upload` (the same
/// endpoint used by chat attachments), so we reuse `AttachmentAPI` instead of inventing
/// a separate uploader.
struct SubmitFeedbackRequest: Encodable {
    var star: Int
    var content: String
    var imagePaths: [String]?

    enum CodingKeys: String, CodingKey {
        case star
        case content
        case imagePaths = "image_paths"
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(star, forKey: .star)
        try c.encode(content, forKey: .content)
        if let imagePaths, !imagePaths.isEmpty {
            try c.encode(imagePaths, forKey: .imagePaths)
        }
    }
}

/// Local attachment state shown in the feedback sheet — same lifecycle pattern as
/// `ChatAttachment`: `.uploading` → `.done(path:)` or `.error(message:)`. Reusing the
/// shared `AttachmentAPI` keeps the upload path consistent across chat + feedback.
struct FeedbackAttachment: Identifiable, Equatable {
    let id: UUID
    let localURL: URL
    let filename: String
    let sizeBytes: Int64
    var status: Status

    init(localURL: URL, filename: String, sizeBytes: Int64, status: Status = .uploading) {
        self.id = UUID()
        self.localURL = localURL
        self.filename = filename
        self.sizeBytes = sizeBytes
        self.status = status
    }

    enum Status: Equatable {
        case uploading
        case done(path: String)
        case error(String)
    }

    var remotePath: String? {
        if case .done(let path) = status { return path }
        return nil
    }

    var isUploading: Bool {
        if case .uploading = status { return true }
        return false
    }

    var displaySize: String {
        let formatter = ByteCountFormatter()
        formatter.allowedUnits = [.useKB, .useMB, .useGB]
        formatter.countStyle = .file
        return formatter.string(fromByteCount: sizeBytes)
    }

    var fileExtension: String {
        (filename as NSString).pathExtension.lowercased()
    }
}

/// Max attachments per submission, matching the web's `maxAttachments` (5).
let FeedbackMaxAttachments: Int = 5

File: apps/qizhi/mobile/MentorAI-Android/app/src/main/kotlin/com/mentorai/app/feedback/FeedbackModels.kt
package com.mentorai.app.feedback

import android.net.Uri
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.Locale
import java.util.UUID

/**
 * Mirrors the server `SubmitFeedbackParams`. `imagePaths` are the storage paths returned by
 * `/attachment/upload` — same endpoint chat uses — so the feedback flow reuses [AttachmentAPI]
 * instead of inventing its own uploader.
 */
@Serializable
data class SubmitFeedbackRequest(
    val star: Int,
    val content: String,
    @SerialName("image_paths") val imagePaths: List<String>? = null,
)

/**
 * Local attachment state shown in the feedback sheet — same lifecycle pattern as
 * `ChatAttachment`: `Uploading → Done(path)` or `Uploading → Error(msg)`.
 */
data class FeedbackAttachment(
    val id: UUID = UUID.randomUUID(),
    val localUri: Uri,
    val filename: String,
    val sizeBytes: Long,
    val status: Status,
) {
    val remotePath: String? get() = (status as? Status.Done)?.path
    val isUploading: Boolean get() = status is Status.Uploading
    val fileExtension: String
        get() = filename.substringAfterLast('.', "").lowercase(Locale.US)

    val displaySize: String get() = formatBytes(sizeBytes)

    sealed class Status {
        object Uploading : Status()
        data class Done(val path: String) : Status()
        data class Error(val message: String) : Status()
    }

    private companion object {
        fun formatBytes(bytes: Long): String {
            if (bytes <= 0) return "0 B"
            val units = arrayOf("B", "KB", "MB", "GB")
            var size = bytes.toDouble()
            var idx = 0
            while (size >= 1024 && idx < units.size - 1) {
                size /= 1024
                idx++
            }
            return String.format(Locale.US, if (size >= 100) "%.0f %s" else "%.1f %s", size, units[idx])
        }
    }
}

/** Max attachments per submission — matches the web's `maxAttachments`. */
const val FEEDBACK_MAX_ATTACHMENTS: Int = 5
